File: python_cpp/uPIMulator_frontend/src/abi/isa/cc/add_nz_cc.py
from typing import Set

from abi.isa.cc._base_cc import BaseCC
from abi.isa.instruction.condition import Condition


class AddNZCC(BaseCC):
    def __init__(self, condition: Condition):
        super().__init__(
            AddNZCC.conditions(), condition,
        )

    @staticmethod
    def conditions() -> Set[Condition]:
        return {
            Condition.C,
            Condition.NC,
            Condition.Z,
            Condition.NZ,
            Condition.XZ,
            Condition.XNZ,
            Condition.OV,
            Condition.NOV,
            Condition.PL,
            Condition.MI,
            Condition.SZ,
            Condition.SNZ,
            Condition.SPL,
            Condition.SMI,
            Condition.NC5,
            Condition.NC6,
            Condition.NC7,
            Condition.NC8,
            Condition.NC9,
            Condition.NC10,
            Condition.NC11,
            Condition.NC12,
            Condition.NC13,
            Condition.NC14,
            Condition.TRUE,
        }

File: golang/uPIMulator/src/linker/kernel/instruction/cc/add_nz_cc.go
package cc

import (
	"errors"
)

type AddNzCc struct {
	condition Condition
}

func (this *AddNzCc) Init(condition Condition) {
	conditions := map[Condition]bool{
		C:    true,
		NC:   true,
		Z:    true,
		NZ:   true,
		XZ:   true,
		XNZ:  true,
		OV:   true,
		NOV:  true,
		PL:   true,
		MI:   true,
		SZ:   true,
		SNZ:  true,
		SPL:  true,
		SMI:  true,
		NC5:  true,
		NC6:  true,
		NC7:  true,
		NC8:  true,
		NC9:  true,
		NC10: true,
		NC11: true,
		NC12: true,
		NC13: true,
		NC14: true,
		TRUE: true,
	}

	if _, found := conditions[condition]; !found {
		err := errors.New("condition is not allowed")
		panic(err)
	}

	this.condition = condition
}

func (this *AddNzCc) Condition() Condition {
	return this.condition
}
